crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs

```diff
@@ -61,13 +64,39 @@ impl Rule for NoConstAssign {
         let model = ctx.model();
 
         let declared_binding = model.declaration(node)?;
-        if let Some(variable_declaration) = declared_binding
-            .syntax()
-            .ancestors()
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
-        {
-            if variable_declaration.is_const() {
-                return Some(declared_binding.syntax().text_trimmed_range());
+
+        if let Some(possible_declarator) = declared_binding.syntax().ancestors().find(|node| {
+            !matches!(
+                node.kind(),
+                JS_OBJECT_BINDING_PATTERN_SHORTHAND_PROPERTY
+                    | JS_OBJECT_BINDING_PATTERN_PROPERTY_LIST
+                    | JS_OBJECT_BINDING_PATTERN_PROPERTY
+                    | JS_OBJECT_BINDING_PATTERN
+                    | JS_ARRAY_BINDING_PATTERN_ELEMENT_LIST
+                    | JS_ARRAY_BINDING_PATTERN
+                    | JS_IDENTIFIER_BINDING
+            )
+        }) {
+            if JsVariableDeclarator::can_cast(possible_declarator.kind()) {
+                let mut possible_declaration = possible_declarator.parent()?;
+                if possible_declaration.kind().is_list() {
+                    possible_declaration = possible_declaration.parent()?;
+                }
+
+                if let Some(js_variable_declaration) =
+                    JsVariableDeclaration::cast_ref(&possible_declaration)
+                {
+                    if js_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
+                if let Some(js_for_variable_declaration) =
+                    JsForVariableDeclaration::cast_ref(&possible_declaration)
+                {
+                    if js_for_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
```

Comment:
I believe we should be able to simplify this because either:

* The parent is a list and the grand parent is the variable declaration
* Or the parent is the `JsForVariableDeclaration`

```suggestion
				if let Some(variable_declaration) = JsForVariableDeclaration::cast_ref(&possible_declaration) {
				  ... 
				} else if let Some(declaration) = JsVariableDeclaratorList::cast_ref(&possible_declaration).and_then(|declaration| declaration.syntax().parent()).and_then(JsVariableDeclaration::cast) {
					.. 
				}
```

Reply:
Because a declaration can be either JsVariableDeclaratorList or JsForVariableDeclaration, I have to do the cast for both types. Your suggestion is to remove the mutable variable, but that would require duplicating the const check.

Reply:
It's correct that you have to test for both cases, but the parent only is a `JsVariableDeclaratorList` if the grandparent is a `JsVariableDeclaration`. The parent of a variable declaration in a for ... in loop is the `JsForVariableDeclaration`. 

You can see this by opening [this playground] and inspecting the syntax tree (you can click on the code on the left and it will jump to the right position in the syntax panel)

Reply:
Thanks @MichaReiser, 
Somehow I thought that the JsVariableDeclaration might be the direct parent of the JsVariableDeclarator. Thank you for explaining all of the details! 
I have updated the PR as you suggested..